### PR TITLE
MQE-2212: Fix invalid behaviour MAGENTO_BACKEND_BASE_URL

### PR DESCRIPTION
**Pull request by soumyau**
### Description

Ported changes from https://github.com/magento/magento2-functional-testing-framework/pull/547 to `develop`

### Fixed Issues (if relevant)

1. magento/magento2-functional-testing-framework#<issue_number>: Issue title
2. ...

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/verification tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)
 - [ ] Changes to Framework doesn't have backward incompatible changes for tests or have related Pull Request with fixes to tests

### CODE REVIEW

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php [hunk 1]
```diff
@@ -2030,20 +2030,23 @@ private function resolveRuntimeReference($args, $regex, $func)
...
-                unset($matches);
-                $replacement = "{$func}(\"{$refVariable}\")";
+                foreach ($matches[0] as $matchKey => $fullMatch) {
+                    $refVariable = $matches[1][$matchKey];
+
+                    $replacement = $this->getReplacement($func, $refVariable);
+
+                    $outputArg = $this->processQuoteBreaks($fullMatch, $newArgs[$key], $replacement);
+                    $newArgs[$key] = $outputArg;
+                }
```
Comment (jilu1) [inline src/Magento/FunctionalTestingFramework/Util/TestGenerator.php:2044]: Why this is changed to a loop?

File: src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php [hunk 2]
```diff
@@ -36,76 +23,61 @@ class MftfGlobals
...
-            . PHP_EOL
-            . 'or'
-            . PHP_EOL
-            . '"MAGENTO_BACKEND_BASE_URL"'
-        );
+        if ($backendBaseUrl) {
+            $backendBaseUrl = UrlFormatter::format($backendBaseUrl, $withTrailingSeparator);
+        }
+        return $backendBaseUrl;
     }
```
[the remark this answers is not in the document]
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php:81]: getBaseUrl('admin') calls this function in the switch case.

File: src/Magento/FunctionalTestingFramework/Util/TestGenerator.php (hunk 1 above)
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Util/TestGenerator.php:2044]: i think this is to accommodate resolving multiple {{_ENV.variable}} references for the fallback function `getReplacement`

File: src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php [hunk 3]
```diff
@@ -36,76 +23,61 @@ class MftfGlobals
...
+                    $baseUrl = self::getBackendBaseUrl() ?: $baseUrl;
+
+                    return UrlFormatter::format(
+                        UrlFormatter::format($baseUrl) . $backendName,
+                        $withTrailingSeparator
+                    );
+            }
+            return UrlFormatter::format($baseUrl, $withTrailingSeparator);
+        } catch (TestFrameworkException $e) {
         }
```
[the remark this answers is not in the document]
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php:51]: $baseUrl changes based on $customArea specified in AdminFormExecutor and FrontendFormExecutor. Caching resulted in failures. How much of a degradation did you see?

File: src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php [hunk 4]
```diff
@@ -4,30 +4,17 @@
...
 use Magento\FunctionalTestingFramework\Exceptions\TestFrameworkException;
+use Magento\FunctionalTestingFramework\Page\Objects\PageObject;
 use Magento\FunctionalTestingFramework\Util\Path\UrlFormatter;
 
 /**
- * MFTF Globals
+ * Provider responsible for returning right URLs for provided scope.
  */
-class MftfGlobals
+class UrlProvider
```
[the remark this answers is not in the document]
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php:16]: I think the reason why UrlProvider is used is to follow the same design pattern as Magento. Refer to `UrlProvider` class in Magento. Not a bad idea to have Url builder class separated out from Global constants. But let me know if you still think if this needs to be refactored to the original.

[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php:16]: Which `UrlProvider` class you refers to? The `Provider` on root level file structure feels misleading.

File: src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php:81]: Does it need to be `public`? I mean if the entry point is `getBaseUrl('admin')`? It seems redundant.

File: src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php:51]: Do you have a build after the change? I remember there were degradation when I make these files change, but I don't remember exactly if it's due to code here.

[the remark this answers is not in the document]
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php:51]: Had linked it in the ticket - https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/34863/

File: src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php:81]: good catch, changed it to private

File: src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php]: Moved `UrlProvider` to Util folder, as per discussion

File: src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php]: I'll trigger a new build once the changes look good.

[the remark this answers is not in the document]
Comment (soumyau) [reply src/Magento/FunctionalTestingFramework/Provider/UrlProvider.php]: 2.4-develop: https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/35087/
MQE-2212: https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/35088/

**Comment by soumyau** [comment]
> * The `Provider` directory is unnecessary. What other providers you expected to belong here?
> * The `baseUrl` and `adminBaseUrl` are two global constants that will not change during test. I don't think we should change to calculate them every time based on `area`. Let's keep the original ideas to keep them cached as private static variables and accessible through getter functions of getBaseUrl() and getBackendBaseUrl() across entire framework. No need to store them separately in `AdminFormExectuor`, `FrontendFormExecutor` and WebApi Executor related classes. The changes should be much simplified. Let me know if you need more details.

Thanks for the comments, AC of this ticket was to port https://github.com/magento/magento2-functional-testing-framework/pull/547 in 3.0. If the AC is not acceptable, it's best to create a new ticket and plan it. What you think?

**Comment by soumyau** [comment]
> * The `Provider` directory is unnecessary. What other providers you expected to belong here?
> * The `baseUrl` and `adminBaseUrl` are two global constants that will not change during test. I don't think we should change to calculate them every time based on `area`. Let's keep the original ideas to keep them cached as private static variables and accessible through getter functions of getBaseUrl() and getBackendBaseUrl() across entire framework. No need to store them separately in `AdminFormExectuor`, `FrontendFormExecutor` and WebApi Executor related classes. The changes should be much simplified. Let me know if you need more details.

I think I understood what you meant by similifying the construction of the url. Updated to code to only pull in relevant pieces from Lukasz's PR just enough to fix the problem.

**Comment by jilu1** [comment]
Changes are simplified and Look good. Make sure to run UR build before merge.

**Comment by soumyau** [comment]
> Changes are simplified and Look good. Make sure to run UR build before merge.

Build looks good - https://m2build-ur.devops.magento.com/job/All-User-Requested-Tests/35159/

**Pull request merged**